### ISSUE TITLE
no-unused-expressions : sequence expressions

### ISSUE THREAD

**Issue by nemoxps** [eslint/eslint#8206]
**Tell us about your environment**

* **ESLint Version:** v3.15.0
* **Node Version:** v7.3.0
* **npm Version:** v3.10.8

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**
.eslintrc.js
```js
module.exports = {
    root: true,
    parserOptions: {
        ecmaVersion: 8,
    },
    rules: {
        'no-unused-expressions': [2, { 'allowShortCircuit': true, 'allowTernary': true }]
    }
};
```

**What did you do? Please include the actual source code causing the issue.**
test.js
```js
let readFile = function (filePath, encoding, cb) {
    if (typeof encoding === 'function')
      cb = encoding, encoding = 'utf8';
    // ...
    return [encoding, cb];
};
```

**What did you expect to happen?**
As the rule name says "no unused expressions", I wouldn't expect an error for the example above.

There is an example in the docs: c = a, b; This is obviously an unused expression and the rule should error in this case. But I can't see something unused in this one: cb = encoding, encoding = 'utf8'; It's just 2 AssignmentExpression inside of a SequenceExpression.
So, I am wondering if I'm right and this rule can be changed slightly.

**What actually happened? Please include the actual, raw output from ESLint.**
$ eslint ./test.js > output.txt

```
C:\nodejs\test\test.js
  3:7  error  Expected an assignment or function call and instead saw an expression  no-unused-expressions

✖ 1 problem (1 error, 0 warnings)
```

**Comment by ljharb** [comment]
`cb = encoding, encoding = 'utf8'` is the same as `cb = (encoding, encoding = 'utf8')`, which is the same as `cb = encoding; encoding = 'utf8';`. I definitely wouldn't expect `no-unused-expressions`, although I'd expect many other rules to warn on the comma operator.

**Comment by platinumazure** [comment]
It's true that expression isn't technically useless, but it is confusing to read and there exists a more conventional equivalent:

```js
cb = encoding;
encoding = "utf8";
```

**Comment by nemoxps** [comment]
As you just stated, as it isn't technically useless, I'd say this rule should allow this syntax. If you'd want to prohibit SequenceExpressions, you can easily use "no-restricted-syntax", can't you?

> ```js
> cb = encoding;
> encoding = "utf8";
I was trying to avoid the curly braces in such a straightforward IfStatement, but I guess you knew that.

**Comment by ljharb** [comment]
I don't think using the comma operator makes anything straightforward.

**Comment by nemoxps** [comment]
Well, I might agree with you, if we add my example from above as an exception.

**Comment by platinumazure** [comment]
Well, for this specific case, could a destructuring assignment work? (This is only feasible if you don't need to support ES5 environments.)

```js
[cb, encoding] = [encoding, "utf8"];
```

**Comment by nemoxps** [comment]
Yes, destructuring would work for my use case; thanks for pointing that out.
So, should we close this issue now, or is it worth getting fixed?

**Comment by ilyavolodin** [comment]
If I remember correctly, when this rule as created, this use case was discussed, and it was decided that sequence expressions are in general very confusing, and should be avoided. So they rule was done this way on purpose (again, if memory serves me right).

**Comment by not-an-aardvark** [comment]
It looks like this behavior is [documented](http://eslint.org/docs/rules/no-unused-expressions#rule-details):

> Sequence expressions (those using a comma, such as `a = 1, b = 2`) are always considered unused unless their return value is assigned or used in a condition evaluation, or a function call is made with the sequence expression value.

So I'll close this since it seems to be working as intended.

**Comment by nemoxps** [comment]
It might be documented, still, as it clearly isn't *useless*, I'd say this is a bug.

**Comment by ljharb** [comment]
If it's deemed that its utility is outweighed by the confusion caused by its use, then the net effect is that it's useless.

**Comment by platinumazure** [comment]
@nemoxps You are more than welcome to create a custom rule that behaves like `no-unused-expressions` except for SequenceExpressions that contain any or all AssignmentExpressions. We just don't see a lot of value in having that behavior in core ESLint at this point. Please feel free to stop by our [Gitter chat](https://gitter.im/eslint/eslint) if you need any help with that.

Thanks again for the suggestion.

**Comment by nemoxps** [comment]
> ```js
> if (typeof encoding === 'function')
>   cb = encoding, encoding = 'utf8';
> ```
If this is confusing to anyone, I'd suggest learning JavaScript properly...

**Comment by ljharb** [comment]
I know JavaScript properly, and I find it confusing. Let's not insult people's intelligence; I could just as equally say that if you find it preferable to write something that *anybody* finds confusing, then you need to learn to program properly - but that wouldn't be polite.

**Comment by nemoxps** [comment]
> if you find it preferable to write something that anybody finds confusing
```js
index => index + 1
```
»Mhh... what's that symbol in the middle? Wow, nice JS, I can write ```>=``` in any order!«

Would you stop using arrow functions (as you "need to learn to program properly") or would you suggest learning the language properly?

**Comment by ljharb** [comment]
This isn't the appropriate place to have this debate.

**Comment by ilyavolodin** [comment]
I'm going to lock this thread. This is not a right venue for this type of conversation.

P.S. @nemoxps You are having a discussion about knowledge of JS with one of the members of TC39. I don' t think the argument is on your side.
